cpp/src/arrow/compute/kernels/vector_cumulative_ops.cc

```diff
@@ -0,0 +1,185 @@
+#include "arrow/compute/kernels/codegen_internal.h"
+#include "arrow/compute/kernels/common.h"
+#include "arrow/result.h"
+#include "arrow/util/bit_util.h"
+#include "arrow/visit_type_inline.h"
+
+namespace arrow {
+namespace compute {
+namespace internal {
+
+// Base class for all cumulative compute functions. Op can be any binary associative
+// operation (+, *, min, etc.) and OptionsType the options type corresponding to Op.
+// ArgType and OutType are the input and output types, which will normally be the
+// same (e.g. the cumulative sum of an array of Int64Type will result in an array of
+// Int64Type).
+template <typename OutType, typename ArgType, typename Op, typename OptionsType>
+struct CumulativeGeneric {
+  using OutValue = typename GetOutputType<OutType>::T;
+  using ArgValue = typename GetViewType<ArgType>::T;
+
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const auto& options = OptionsWrapper<OptionsType>::Get(ctx);
+    auto skip_nulls = options.skip_nulls;
+
+    auto out_type = batch.GetDescriptors()[0].type;
+    ARROW_ASSIGN_OR_RAISE(
+        auto cast_value,
+        Cast(Datum(options.start), out_type, CastOptions::Safe(), ctx->exec_context()));
+    auto start = UnboxScalar<OutType>::Unbox(*cast_value.scalar());
+
+    int64_t base_output_offset = 0;
+    bool encountered_null = false;
+    ArrayData* out_arr = out->mutable_array();
+
+    switch (batch[0].kind()) {
+      case Datum::ARRAY: {
+        auto st = Call(ctx, base_output_offset, *batch[0].array(), out_arr, &start,
+                       skip_nulls, &encountered_null);
+        out_arr->SetNullCount(arrow::kUnknownNullCount);
+        return st;
+      }
+      case Datum::CHUNKED_ARRAY: {
+        const auto& input = batch[0].chunked_array();
+
+        for (const auto& chunk : input->chunks()) {
+          RETURN_NOT_OK(Call(ctx, base_output_offset, *chunk->data(), out_arr, &start,
+                             skip_nulls, &encountered_null));
+          base_output_offset += chunk->length();
+        }
+        out_arr->SetNullCount(arrow::kUnknownNullCount);
+        return Status::OK();
+      }
+      default:
+        return Status::NotImplemented(
+            "Unsupported input type for function 'cumulative_<operator>': ",
+            batch[0].ToString());
+    }
+  }
+
+  static Status Call(KernelContext* ctx, int64_t base_output_offset,
+                     const ArrayData& input, ArrayData* output, ArgValue* accumulator,
+                     bool skip_nulls, bool* encountered_null) {
+    Status st = Status::OK();
+    auto out_bitmap = output->GetMutableValues<uint8_t>(0);
+    auto out_data = output->GetMutableValues<OutValue>(1) + base_output_offset;
+    int64_t curr = base_output_offset;
+
+    auto null_func = [&]() {
+      *out_data++ = OutValue{};
+      *encountered_null = true;
+      arrow::bit_util::SetBitTo(out_bitmap, curr, false);
```

Comment:
Do we know why null-handling with `INTERSECTION` is not working correctly for `ChunkedArrays`?

Reply:
See ARROW-16145